### PR TITLE
Checker config

### PR DESCRIPTION
**Pull request by bruntib**

### CODE REVIEW

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
- analyzer/codechecker_analyzer/cmd/analyze.py  2
         

Complexity decreasing per file
==============================
+ analyzer/codechecker_analyzer/cmd/checkers.py  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=5365850&bid=17582932)

File: analyzer/codechecker_analyzer/cmd/checkers.py [hunk 1]
```diff
@@ -190,68 +181,73 @@ def main(args):
...
+        context)
 
     analyzer_environment = env.extend(context.path_env_extra,
                                       context.ld_lib_path_extra)
 
     analyzer_config_map = analyzer_types.build_config_handlers(args,
                                                                context,
                                                                working)
+
+    def uglify(text):
```
Comment (csordasmarton) [inline analyzer/codechecker_analyzer/cmd/checkers.py:199]: Function docstring is missing.

File: docs/analyzer/user_guide.md [hunk 2]
```diff
@@ -717,6 +717,11 @@ analyzer arguments:
                         clang-tidy checkers. The file can be dumped by
                         'CodeChecker analyzers --dump-config clang-tidy'
                         command.
+  --checker-config [CHECKER_CONFIG [CHECKER_CONFIG ...]]
+                        Checker configuration options in the following format:
+                        analyzer:key=value. The collection of the options can
+                        be printed with 'CodeChecker checkers
```
Comment (csordasmarton) [inline docs/analyzer/user_guide.md:723]: If I run the following command: `CodeChecker checkers --checker-config` I will get the following error message:
```sh
Traceback (most recent call last):
  File "~/codechecker/build/CodeChecker/cc_bin/CodeChecker.py", line 128, in main
    args.func(args)
  File "~/codechecker/build/CodeChecker/cc_bin/../lib/python3/codechecker_analyzer/cmd/checkers.py", line 197, in main
    working)
  File "~/codechecker/build/CodeChecker/lib/python3/codechecker_analyzer/analyzers/analyzer_types.py", line 203, in build_config_handlers
    construct_config_handler(args, context)
  File "~/codechecker/build/CodeChecker/lib/python3/codechecker_analyzer/analyzers/clangtidy/analyzer.py", line 292, in construct_config_handler
    for cfg in args.checker_config:
TypeError: 'bool' object is not iterable
```

File: docs/analyzer/user_guide.md [hunk 3]
```diff
@@ -1166,6 +1171,8 @@ optional arguments:
                         List checkers enabled by the selected profile.
                         'list' is a special option showing details about
                         profiles collectively.
+  --checker-config      Show checker configuration options. These can be
+                        given to 'CodeChecker analyze --checker-config'.
```
Comment (csordasmarton) [inline docs/analyzer/user_guide.md:1175]: It can be given to the `CodeChecker check` command too.

File: docs/analyzer/user_guide.md [hunk 4]
```diff
@@ -717,6 +717,11 @@ analyzer arguments:
                         clang-tidy checkers. The file can be dumped by
                         'CodeChecker analyzers --dump-config clang-tidy'
                         command.
+  --checker-config [CHECKER_CONFIG [CHECKER_CONFIG ...]]
```
Comment (csordasmarton) [inline docs/analyzer/user_guide.md:720]: Please extend the user guide for the check command too: https://github.com/Ericsson/codechecker/blob/master/docs/analyzer/user_guide.md#check-

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- analyzer/codechecker_analyzer/cmd/analyze.py  3
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
         

Complexity decreasing per file
==============================
+ analyzer/codechecker_analyzer/cmd/checkers.py  -3
+ analyzer/codechecker_analyzer/cmd/check.py  -1
         

Clones removed
==============
+ analyzer/codechecker_analyzer/cmd/analyze.py  -1
+ analyzer/codechecker_analyzer/cmd/check.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=5365850&bid=17582932)

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- analyzer/codechecker_analyzer/cmd/analyze.py  3
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
         

Complexity decreasing per file
==============================
+ analyzer/codechecker_analyzer/cmd/checkers.py  -3
+ analyzer/codechecker_analyzer/cmd/check.py  -1
         

Clones removed
==============
+ analyzer/codechecker_analyzer/cmd/analyze.py  -1
+ analyzer/codechecker_analyzer/cmd/check.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=5365850&bid=17582932)

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py  5
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
- analyzer/codechecker_analyzer/cmd/analyzers.py  2
- analyzer/codechecker_analyzer/cmd/analyze.py  6
         

Complexity decreasing per file
==============================
+ analyzer/codechecker_analyzer/cmd/checkers.py  -3
+ analyzer/codechecker_analyzer/cmd/check.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=5365850&bid=17582932)

File: analyzer/codechecker_analyzer/cmd/analyze.py [hunk 5]
```diff
@@ -302,6 +292,27 @@ def add_arguments_to_parser(parser):
                                     "'CodeChecker analyzers --dump-config "
                                     "clang-tidy' command.")
 
+    analyzer_opts.add_argument('--analyzer-config',
+                               dest='analyzer_config',
+                               nargs='*',
+                               default=argparse.SUPPRESS,
```
Comment (dkrupp) [inline analyzer/codechecker_analyzer/cmd/analyze.py:298]: We should go with the default for clang-tidy to show findings in headers (but not from system headers).
so the default should be:
--analyze-config clang-tidy:HeaderFilterRegex=*

and we should add an explicit option to take all configuration from .tidy-config files:
clang-tidy:take-config-from-directory


We should describe all this config behaviour  in the analyzer user guide.

Comment (bruntib) [reply analyzer/codechecker_analyzer/cmd/analyze.py:298]: I think this should be done in #2690, because this is another issue.

File: analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py [hunk 6]
```diff
@@ -379,4 +395,29 @@ def construct_config_handler(cls, args, context):
             cmdline_checkers,
             'enable_all' in args and args.enable_all)
 
+        handler.checker_config = []
+        r = re.compile(r'(?P<analyzer>.+?):(?P<key>.+?)=(?P<value>.+)')
+
+        # TODO: This extra "isinstance" check is needed for
+        # CodeChecker checkers --checker-config. This command also runs
+        # this function in order to construct a config handler.
+        if 'checker_config' in args and \
```
Comment (csordasmarton) [inline analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py:404]: If this option is given in the command line it will be always a list. If you remove the `default=argparse.SUPPRESS` from the argparse you can change this condition to: `if args.checker_config:`.

File: analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py [hunk 7]
```diff
@@ -379,4 +395,29 @@ def construct_config_handler(cls, args, context):
...
+            for cfg in args.checker_config:
+                m = re.search(r, cfg)
+                if m.group('analyzer') == cls.ANALYZER_NAME:
+                    handler.checker_config.append(
+                        m.group('key') + '=' + m.group('value'))
+
+        # TODO: This extra "isinstance" check is needed for
+        # CodeChecker analyzers --analyzer-config. This command also runs
+        # this function in order to construct a config handler.
+        if 'analyzer_config' in args and \
```
Comment (csordasmarton) [inline analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py:415]: Same as above.

File: analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py [hunk 8]
```diff
@@ -267,19 +308,44 @@ def construct_config_handler(cls, args, context):
...
+
+            for cfg in args.checker_config:
+                m = re.search(r, cfg)
+                if m.group('analyzer') == cls.ANALYZER_NAME:
+                    check_options.append({'key': m.group('key'),
+                                          'value': m.group('value')})
+            analyzer_config['CheckOptions'] = check_options
+        else:
+            try:
+                with open(args.tidy_config, 'r') as tidy_config:
```
Comment (csordasmarton) [inline analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py:339]: The following arguments are missing from the open call: `encoding="utf-8", errors="ignore"`

File: analyzer/codechecker_analyzer/cmd/analyze.py [hunk 9]
```diff
@@ -238,17 +239,6 @@ def add_arguments_to_parser(parser):
                                     ', '.join(analyzer_types.
                                               supported_analyzers) + ".")
 
-    analyzer_opts.add_argument('--add-compiler-defaults',
```
Comment (csordasmarton) [inline analyzer/codechecker_analyzer/cmd/analyze.py:241]: @gyorb When do we remove a deprecated option? Do we need to wait some releases or what is the process of it?

File: analyzer/codechecker_analyzer/cmd/analyze.py [hunk 10]
```diff
@@ -693,6 +704,23 @@ def main(args):
...
     LOG.debug("Output will be stored to: '" + args.output_path + "'")
 
+    config_option_re = re.compile(r'^({}):.+=.+$'.format(
+        '|'.join(analyzer_types.supported_analyzers)))
+
+    # Check the format of analyzer options.
+    if 'analyzer_config' in args:
+        for config in args.analyzer_config:
+            if not re.match(config_option_re, config):
+                LOG.error("Analyzer option in wrong format: " + config)
```
Comment (csordasmarton) [inline analyzer/codechecker_analyzer/cmd/analyze.py:714]:
```suggestion
                LOG.error("Analyzer option in wrong format: %s", config)
```

File: analyzer/codechecker_analyzer/cmd/analyze.py [hunk 11]
```diff
@@ -693,6 +704,23 @@ def main(args):
...
+                sys.exit(1)
+
+    # Check the format of checker options.
+    if 'checker_config' in args:
+        for config in args.checker_config:
+            if not re.match(config_option_re, config):
+                LOG.error("Checker option in wrong format: " + config)
```
Comment (csordasmarton) [inline analyzer/codechecker_analyzer/cmd/analyze.py:721]:
```suggestion
                LOG.error("Checker option in wrong format: %s", config)
```

File: analyzer/codechecker_analyzer/cmd/checkers.py [hunk 12]
```diff
@@ -61,33 +59,18 @@ def get_warnings(env=None):
...
             universal_newlines=True,
             encoding="utf-8",
             errors="ignore")
-        return parse_warnings(result)
+        return result.split()
     except (subprocess.CalledProcessError, OSError):
         return []
 
 
-def parse_warnings(diagtool_output):
```
Comment (csordasmarton) [inline analyzer/codechecker_analyzer/cmd/checkers.py:77]: Why did you remove this function? Without this it will not skip some line from the diagtool output:
```

GREEN = enabled by default

-W
  -Wextra
```
This function also removed the white spaces.

File: analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py (hunk 6 above)
Comment (bruntib) [reply analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py:404]: Yes, this was my original solution. However, unfortunately this `construct_config_handler()` function is called when `CodeChecker checkers --checker-config` command is run and in that sub-command this flag has no argument. In this case the value of `args.checker_config` is `True`. After fixing the TODO above this can also be fixed.

File: analyzer/codechecker_analyzer/cmd/checkers.py (hunk 12 above)
Comment (bruntib) [reply analyzer/codechecker_analyzer/cmd/checkers.py:77]: The first line goes to the standard error, so it is not in the standard output what we parse. The extra spaces are eliminated by `split()` function.

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
- analyzer/codechecker_analyzer/cmd/analyze.py  6
- analyzer/codechecker_analyzer/cmd/analyzers.py  4
- analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py  5
         

Complexity decreasing per file
==============================
+ analyzer/codechecker_analyzer/cmd/check.py  -1
+ analyzer/codechecker_analyzer/cmd/checkers.py  -1
         

Clones added
============
- analyzer/tests/functional/analyze/test_analyze.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=5365850&bid=17582932)

File: analyzer/tests/functional/analyze/test_analyze.py [hunk 13]
```diff
@@ -712,3 +712,50 @@ def test_multiple_invalid_checker_names(self):
 
         errcode = process.returncode
         self.assertEqual(errcode, 0)
+
+    def test_analyzer_and_checker_config(self):
```
Comment (csordasmarton) [inline analyzer/tests/functional/analyze/test_analyze.py:716]: Docstring is missing.

File: analyzer/tests/functional/analyze/test_analyze.py [hunk 14]
```diff
@@ -712,3 +712,50 @@ def test_multiple_invalid_checker_names(self):
...
+        build_log = [{"directory": self.test_workspace,
+                      "command": "gcc -c " + source_file,
+                      "file": source_file}]
+
+        with open(build_json, 'w',
+                  encoding="utf-8", errors="ignore") as outfile:
+            json.dump(build_log, outfile)
+
+        analyze_cmd = [self._codechecker_cmd, "check", "-l", build_json,
+                       "--analyzers", "clang-tidy", "-o", self.report_dir,
```
Comment (csordasmarton) [inline analyzer/tests/functional/analyze/test_analyze.py:728]: Could you please add a new test case or extend this one and enable clangsa analyzer and add some analyzer and checker config for the analyze command to make sure that it works properly with this analyzer too.

File: docs/analyzer/user_guide.md [hunk 15]
```diff
@@ -219,6 +222,16 @@ analyzer arguments:
                         clang-tidy checkers. The file can be dumped by
                         'CodeChecker analyzers --dump-config clang-tidy'
                         command.
+  --analyzer-config [ANALYZER_CONFIG [ANALYZER_CONFIG ...]]
+                        Analyzer configuration options in the following format:
+                        analyzer:key=value. The collection of the options can
+                        be printed with 'CodeChecker analyzers
+                        --analyzer-config'.
+  --checker-config [CHECKER_CONFIG [CHECKER_CONFIG ...]]
```
[the remark this answers is not in the document]
Comment (csordasmarton) [reply docs/analyzer/user_guide.md:232]: You marked this as resolved but it still does not work for me.

File: docs/analyzer/user_guide.md (hunk 3 above)
Comment (csordasmarton) [reply docs/analyzer/user_guide.md:1240]: You marked this as resolved but you didn't changed it.

File: docs/analyzer/user_guide.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (bruntib) [reply docs/analyzer/user_guide.md:232]: It works for me. Am I missing something?

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
- analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py  5
- analyzer/codechecker_analyzer/cmd/analyzers.py  4
- analyzer/codechecker_analyzer/cmd/analyze.py  6
         

Complexity decreasing per file
==============================
+ analyzer/codechecker_analyzer/cmd/check.py  -1
+ analyzer/codechecker_analyzer/cmd/checkers.py  -1
         

Clones added
============
- analyzer/tests/functional/analyze/test_analyze.py  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=5365850&bid=17582932)

File: docs/analyzer/user_guide.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (csordasmarton) [reply docs/analyzer/user_guide.md:232]: If I run the `CodeChecker checkers --checker-config` command it will print the checker configs:
```
 clangsa:alpha.clone.CloneChecker:MinimumCloneComplexity
 clangsa:alpha.clone.CloneChecker:ReportNormalClones
 clangsa:alpha.clone.CloneChecker:IgnoredFilesPattern
 clangsa:alpha.security.MmapWriteExec:MmapProtExec
 clangsa:alpha.security.MmapWriteExec:MmapProtRead
 clangsa:cplusplus.Move:WarnOn
 ...
```
Which is good, but if I run the `CodeChecker checkers --detail` command it will not print these options below each checkers:
```sh
 - alpha.clone.CloneChecker                                       clangsa    LOW         (Enable only for development!) Reports similar pieces of code.
 - alpha.core.BoolAssignment                                      clangsa    LOW         (Enable only for development!) Warn about assigning non-{0,1} values to Boolean variables
```

The expected output of this command would be the following:
```
- alpha.clone.CloneChecker                                       clangsa    LOW         (Enable only for development!) Reports similar pieces of code.
  Options:
  clangsa:alpha.clone.CloneChecker:MinimumCloneComplexity
  clangsa:alpha.clone.CloneChecker:ReportNormalClones
  clangsa:alpha.clone.CloneChecker:IgnoredFilesPattern
```

**Comment by csordasmarton** [comment]
@bruntib Please resolve the conflict.

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py  5
- analyzer/codechecker_analyzer/cmd/analyzers.py  4
- analyzer/codechecker_analyzer/cmd/analyze.py  6
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
         

Complexity decreasing per file
==============================
+ analyzer/codechecker_analyzer/cmd/check.py  -1
+ analyzer/codechecker_analyzer/cmd/checkers.py  -1
         

Clones added
============
- analyzer/tests/functional/analyze/test_analyze.py  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=5365850&bid=17582932)
